CompoundEditor : Redesign pinning menu

themissingcow opened the pull request:
![image](https://user-images.githubusercontent.com/896779/66818540-60ed9980-ef35-11e9-98b2-e9aa215fb7d2.png)

### Background

The pinning options in Gaffer have evolved significantly since 0.53. To the extent that it's now somewhat confusing. The old UI we had doesn't really work very well now you have these other modes available. For reference, as of 55 compared to 53, you will have:

 - Editor linking
 - Additional follow modes (eg: edit source node for selection, numeric bookmarks) 
 - The ability to save linking/following state as part of a layout

There has also been discussion around it being difficult to tell the state of an editor at a glance.   As such, for 0.55 so we can conclude the series of pinning workflow tweaks, we're attempting to make all of these new options not only easier to find, but hopefully make a bit more sense.


### User facing changes

The pinning button has been completely re-designed and replaced by the Editor Focus menu. All functionality is accessed via the left mouse button.

The menu visually reflects the current focus mode for the editor with an icon and which editor group it belongs to by its color.

It allows a user to choose between several editor focus modes, falling into two categories:

 - Follow: The editor will update which node(s) it operates on based on some logic; the selected node(s), the node assigned to a bookmark, etc.

 - Pin: The editor will operate on fixed node(s) and will not change unless another focus mode is selected.

### API changes

API Changes:

 - NodeSetEditor.registerNodeSetDriverMode: Adds optional `description` kwarg to supply a custom UI description for the mode.

 - Adds `NodeSetEditor.nodeSetDriverModeDescription` to retrieve the UI description for a specific driver mode.

 - `Gaffer.Menu` now supports specifying a `label` when `divider` is True. If provided, the simple line divider will be replaced with a visually distinct text heading containing the label.

themissingcow commented:
@medubelko - be great to get your thoughts on this. We are aiming to release this beginning of next week I think.

I also attempted to fix-up the current docs to reflect the changes to the menu in https://github.com/GafferHQ/gaffer/pull/3399/commits/af2a63f542351cb05d16d0119828f273b411c593, would greatly appreciate your discerning eye. What are you thoughts more generally on what else we might want to do/change?

themissingcow commented:
Thanks @andrewkaufman 

> * Are we sold on G for unpin/unfollow? I guess it stands for Global? It was a bit unintuitive... or maybe my brain was primmed for U...

We're trying to move away from 'unpinned', as 'unpinned' is really 'Following Node Selection' in new speak. I just realised the other day that `g` is useless too as it switches to the green channel in an ImageView 🙃 . I've made in `n` for 'Node Selection`. How about that?

> * Should P/G on a slave editor be breaking the link or updating the driver (like drag/drop does)? I was leaning towards the later but I could be convinced otherwise.

There is a bit of a dilemma here. It feels a lot more useful for things like drag/numbers to update the master editor.

These two are, however, listed in the menu itself as shortcut keys. This makes the user think that pressing that key is the same as choosing that menu item. Using the menu would break the link as it has to as its the only way to do it.

So seems like we have 2 options:

 - Dont list the shortcuts in the menu, so they're not associated with the menu item, (as its only `p` and `n` that are), and describe them as shortcuts to update the driver.

 - Never update the driver for drag/drop or keyboard actions (unless you perform them on the driver of course), which makes things more consistent.

> * When the master is following a numeric bookmark we get a nice ⭐️ `2` icon, but the slaves remain a 🔗, which is a bit confusing. I think people are really going to want to see that ⭐️ `2` everywhere. The color is enough to know its linked.

It's not just numeric bookmarks, whatever the icon is for the mode, you only see the link icon if the editor is a slave.    I'd fear it was too hard to tell which was the master (with the current design) if we remove the link icon.   The only alternative I can think of is to totally re-design it back again the other way so linking is totally separate from the focus mode itself and you have 2 widgets.

johnhaddon commented:
> Should P/G on a slave editor be breaking the link or updating the driver (like drag/drop does)? I was leaning towards the later but I could be convinced otherwise.

I'm unsure that drag/drop should be updating the driver in the first place. Drag and drop is an interaction with a very specific source and destination. If I go to the trouble of dragging a node specifically into the SceneInspector, I don't expect the Viewer to change. And the Viewer itself is such a prominent part of the UI that I'm not sure there is much convenience to be had in using other editors as a proxy for it.

> When the master is following a numeric bookmark we get a nice star 2 icon, but the slaves remain a link, which is a bit confusing.

I think it's important for the master/slave relationship to be visible at a glance somehow, thus don't want to lose the link icon.

themissingcow commented:
@johnhaddon @andrewkaufman For now, in https://github.com/GafferHQ/gaffer/commit/f10d4d9b46231de6193c309dec0eef1b7ccd5177 I've removed the shortcuts from the menu, and updated the code to affect the master editor - simply so its consistent with other shortcuts.

More than happy to make another commit that updates drag/drop and all editor focus shortcuts to break the link instead and only affect the editor receiving the key event.

johnhaddon commented:
> More than happy to make another commit that updates drag/drop and all editor focus shortcuts to break the link instead and only affect the editor receiving the key event.

For this PR I think it's most important that we make all the actions consistent, so where we're at now is OK - I'm just waiting for Azure before merging. This stuff is a huge improvement over what we had before so I don't want to hold it up on a tangential discussion.

But I do think a drag affecting something other than the destination is unexpected and unhelpful, so Tom's gonna do some impartial user research to see if I'm alone in that. If necessary, we'll deal with the outcome of that later...

themissingcow commented:
I just noticed, somewhere in here I've broken node drag/drop. Don't merge yet.

themissingcow commented:
> I just noticed, somewhere in here I've broken node drag/drop. Don't merge yet.

Sorry for the delay there - inadvertent 'one `dd` to many in `vim`. Fixed in https://github.com/GafferHQ/gaffer/pull/3399/commits/e1b756168a425e2cd57300c5f7255a5dca6ef538.

andrewkaufman commented:
> I've made in n for 'Node Selection`. How about that?

Sure, lets go for it.

> I'm unsure that drag/drop should be updating the driver in the first place. Drag and drop is an interaction with a very specific source and destination. If I go to the trouble of dragging a node specifically into the SceneInspector, I don't expect the Viewer to change.

Another way to phrase that same sentiment could be that _your_ default layout should have an unlinked SceneInspector.

For me, I might agree with you on the SceneInspector, but I definitely wouldn't on the HierarchyView. I guess the key difference is both the Viewer & HierarchyView are thought of as "drivers", so I instinctively use them interchangeably to drive all linked editors, and I go for whichever one my mouse is closest to.

> you only see the link icon if the editor is a slave. I'd fear it was too hard to tell which was the master

> I think it's important for the master/slave relationship to be visible at a glance somehow, thus don't want to lose the link icon.

For me, its more important to know what data-source (node) each editor is displaying at a glance than it is to know whether its a slave editor. Currently users have to glance, see a :link:, see :red_circle: , then make several glances to find the :red_circle: :star: `2` to finally find out what node they're inspecting/editing.

> But I do think a drag affecting something other than the destination is unexpected and unhelpful, so Tom's gonna do some impartial user research to see if I'm alone in that. If necessary, we'll deal with the outcome of that later...

If you're going to do impartial user research, I'd be keen to know how many users have use cases that actually require master/slave relationships. I was totally thrown by it when the PR first went up, and personally have no need for it. For me, linking can be entirely commutative. I wouldn't link editors unless I wanted them to all update in unison, if I do link editors I'd wouldn't want them to be easily/mistakenly unlinked, and if/when I want them unlinked I would be happy to explicitly unlink them using the menu.

> For this PR ...where we're at now is OK. This stuff is a huge improvement over what we had before so I don't want to hold it up on a tangential discussion.

Totally agree with this sentiment. Its a great improvement already. I'd be keen to roll it out as is, do some casual user surveying over the course of a month or two, then circle back with some updates if necessary.

johnhaddon commented:
> Another way to phrase that same sentiment could be that your default layout should have an unlinked SceneInspector.

The sentiment I expressed was a general one, about the nature of drag and drop operations, and the expected target of the operation being the destination of the drag. I don't see how my choice of default layout relates to that. My preferred layout is definitely to have the inspectors follow the viewer by default.

> I go for whichever one my mouse is closest to.

Your argument seems to be largely convenience based (at least until linking is commutative - see below). With the current behaviour, to pin the SceneInspector, you go to the SceneInspector, open the menu, unlink it, go back to the GraphEditor and then drag the node into the SceneInspector. That's not particularly convenient. On the other hand, both the Viewer and SceneHierarchy are usually front and centre and easy to hit - they don't need to be proxies for each other.

One thing that I'd like though is the ability to easily make floating inspectors for the current node/object, similar to how you can make floating NodeEditors easily. If we had that I wonder how much I'd want to muck about with individual pinning anyway.

> For me, linking can be entirely commutative.

I think this is perhaps the crux of it. If linking was presented that way, then it's more reasonable to think of the drag target as being all the linked editors, not just the one the mouse ends up in. But it's not presented that way currently. I seem to remember wanting things to be more commutative at the API level as well, but forget why we didn't go that way. It's something I would like to revisit when we refactor Editors to be GraphComponents (which we want to do so we can serialise all of their settings).

> I'd be keen to roll it out as is, do some casual user surveying over the course of a month or two, then circle back with some updates if necessary.

Yep, that's the plan.

themissingcow commented:
@andrewkaufman Do you want to have a quick chat next week, which may be easier than typing?

re: 

>It's not just numeric bookmarks, whatever the icon is for the mode, you only see the link icon if the editor is a slave.

I was talking rubbish there. Sorry, long day. Ignore me.

I think John has it here:

>>For me, linking can be entirely commutative.
>I think this is perhaps the crux of it.

as to

>I seem to remember wanting things to be more commutative at the API level as well, but forget why we didn't go that way.

iirc, it was that for 'basic' linking, being commutative makes sense. But take the assorted 'other' link modes. Eg: Source node for Scene Selection. That by definition can't be. It's 'this editor shows the source node for the selection, based on the node viewed by the linked editor'. The master editor is likely not showing the same node as the slave in this case..

There are other follow modes that we want to support in the future (people are asking about follow tweak for scene selection, viewer auto-lookthorugh, etc...). So thinking of linking in its simplistic terms to inform the UI here, probably only covers a smaller subset of the eventual use cases.

I think either way there are a couple of mental models people operate under that we need to support, so maybe a chat is best?  For example, in a quick poll at CS, no-one wanted the hotkeys to update the master editor, everyone who voted wanted link breaking...

themissingcow commented:
> ...then make several glances to find the 🔴 ⭐️ `2` to finally find out what node they're inspecting/editing

Are you explicitly wanting to know exactly which node they are editing, or instead wanting to know the 'rule' for the node they are editing? (Ie: 'ShaderAssignment21' vs 'Following Bookmark 2'). Just that some editors show you the target node on screen, some don't, which might need some refinement.

> If you're going to do impartial user research, I'd be keen to know how many users have use cases that actually require master/slave relationships. I was totally thrown by it when the PR first went up, and personally have no need for it. For me, linking can be entirely commutative. I wouldn't link editors unless I wanted them to all update in unison

Basic editor linking was actually the second use case, behind one that was explicitly a master/slave scenario. The entire topic came up to support 'constant `alt-e` mode' for the Node Editor, which people have been very excited about when we p-o-c'd it.  That only makes sense in a master/slave mode. There are further extensions to this too, people are already asking for a 'follow tweak node' focus mode too, which again by definition isn't commutative.

We also had a lot of feedback from people in our travels that the find managing pining cumbersome and confusing, especially from new and/or less technical artists.  It was one of the main overview points from my research at CS was that "people have to spend too much time managing and jumping between tabs and panels". Another very common request was for Shake/Katana/Houdini style global view/edit flags (so multiple editors could be updated in once go). As such, given we had a working mechanism for the concept of the `alt-e` follow, it felt like a logical extension to add a simple literal follow as well, to help those who found it very confusing that the viewer/hierarchy view _werent_ linked in the first place. Then at least there was one way to manage the focus mode, that supported both commutative and non-commutative modes. 

> > For this PR ...where we're at now is OK. This stuff is a huge improvement over what we had before so I don't want to hold it up on a tangential discussion.
> 
> Totally agree with this sentiment. Its a great improvement already. I'd be keen to roll it out as is, do some casual user surveying over the course of a month or two, then circle back with some updates if necessary.

Grand, thanks. Definitely something for refinement over time. 

There is another area that would be very good to check in on too. Initially All inspectors were linked to the Hierarchy View, and that was linked to the Viewer.  The intention was to make it easy to de-couple the viewer (eg: pin-off), but there was a little confusion there, and we also wanted to try and move people away from pinning of the viewer, and to using pause instead (as it was a nasty workaround anyway). In use, there may be more cases where people want to pin the viewer to a 2D node, and easily keep the rest of the 3D editors linked, so going back to `viewer < hierarchy view < others` may make more sense. Or something else entirely?

I still don't really like most of the icons either, but have tried a bunch of other options, so if anyone has any ideas there...)

John mentioned last week to merge the PR if CI passed, so if you're happy as-is too then I'll merge this and we can refine as we go.

andrewkaufman commented:
> Are you explicitly wanting to know exactly which node they are editing, or instead wanting to know the 'rule' for the node they are editing? (Ie: 'ShaderAssignment21' vs 'Following Bookmark 2').

A bit of both. I want to know the rule, but I also want to use the numeric bookmarks to imply a more meaningful name. eg. 'ImageMerge27' vs 'Bookmark 2 which I know means IPR plus slap comp`. I think we're probably going to be requesting named rather than numbered bookmarks at some point (or at least ability to label the numbered ones with custom text).

> Basic editor linking was actually the second use case, behind one that was explicitly a master/slave scenario. The entire topic came up to support 'constant alt-e mode' for the Node Editor, which people have been very excited about when we p-o-c'd it. 

It was not. You're only referring to one group of users. Another group of users requested the basic linking long before the alt-e mode existed.

> There are further extensions to this too, people are already asking for a 'follow tweak node' focus mode too, which again by definition isn't commutative.

Which Tweak node? The last one? The one that's tweaking exposure? I'm not sure the master/slave following is well suited to solve this user request. It seems like it'd be better solved with the EditScopes, spreadsheet node, viewport light manipulators, the uber lighting panel, and/or your ephemeral/transcient plug aggregation hud idea.

To be honest, I think the same of alt+e source edits... I don't often come across scripts where source nodes are actually editable... a lot of people work with most/all sources locked (references, external cache files) and aggregation of various tweaks scattered all over the graph are more relevant (hence the request for following tweaks).

> Then at least there was one way to manage the focus mode, that supported both commutative and non-commutative modes.

In my opinion, this is at the expense of the more useful commutative mode. But maybe its only more useful for me. We can wait and see, and I can live with it if that's the case.

themissingcow commented:
> > Basic editor linking was actually the second use case, behind one that was explicitly a master/slave scenario. The entire topic came up to support 'constant alt-e mode' for the Node Editor, which people have been very excited about when we p-o-c'd it.
> 
> It was not. You're only referring to one group of users. Another group of users requested the basic linking long before the alt-e mode existed.

Haha, fair enough, I'd never seen that request... I just added it in as it seemed like it would be useful once we'd done the alt-e mode...
 
> > There are further extensions to this too, people are already asking for a 'follow tweak node' focus mode too, which again by definition isn't commutative.
> 
> Which Tweak node? The last one? The one that's tweaking exposure?

Starter for 10 would be the same behaviour as `alt-shift-e`...

> I'm not sure the master/slave following is well suited to solve this user request. It seems like it'd be better solved with the EditScopes, spreadsheet node, viewport light manipulators, the uber lighting panel, and/or your ephemeral/transcient plug aggregation hud idea.

Yeah, these all will definitely evolve as that becomes available.

> To be honest, I think the same of alt+e source edits... I don't often come across scripts where source nodes are actually editable... a lot of people work with most/all sources locked (references, external cache files) and aggregation of various tweaks scattered all over the graph are more relevant (hence the request for following tweaks).

I think this just highlights the diversity of Gaffer use cases now, FA, and VFX in London, where artists are directly creating lights definitely have editable ones.

johnhaddon commented:
> It was not. You're only referring to one group of users. Another group of users requested the basic linking long before the alt-e mode existed.

History aside, we do now need to support both things.

> for 'basic' linking, being commutative makes sense. But take the assorted 'other' link modes. Eg: Source node for Scene Selection. That by definition can't be. It's 'this editor shows the source node for the selection, based on the node viewed by the linked editor'.

For cummutativity, does it work to phrase that as "this editor is grouped with the viewer et al, but is additionally in 'follow scene selection mode'"? I guess I still can't quite picture what the menus look like if we make things commutative though. Are there  "Join Group" and "Leave Group" items? And do all other items act on the entire group as a whole?

> There is another area that would be very good to check in on too. Initially All inspectors were linked to the Hierarchy View, and that was linked to the Viewer. The intention was to make it easy to de-couple the viewer

I initially found this confusing, so its down to me that we backed out and just linked everything direct to the Viewer. I'm not 100% sure I made the right call though : pinning an _image_ into the viewer now renders the scene editors inoperable. Perhaps two viewers is the answer to that, but I definitely don't want to rewrite everything for commutative linking until we have a definitive answer there.

> We can wait and see

At least we're all on the same page with this wait-and-see approach. Tom's made huge improvements to the editor focus stuff, and while there may be open questions, I don't see any reason not to get those improvements into people's hands and seek answers from there. If after all that, you still want to champion commutative linking, I'd suggest making a proposal that shows how the existing modes can be incorporated into it and what the menus should look like...
